README.md
# SWIFT MT940 bank statement format JS parser

[![Build Status](https://travis-ci.org/a-fas/mt940js.svg?branch=master)](https://travis-ci.org/a-fas/mt940js)
[![NPM version](https://badge.fury.io/js/mt940js.svg)](https://badge.fury.io/js/mt940js)
[![codecov](https://codecov.io/gh/a-fas/mt940js/branch/master/graph/badge.svg)](https://codecov.io/gh/a-fas/mt940js)

*[History of changes](./CHANGELOG)*

mt940js is a SWIFT mt940 bank statement format parser for javascript (ES2015). Takes in text of mt940 file, puts out array of parsed statements and transactions.
See examples below.

## Installation

```bash
npm install mt940js
```

Also accessible via unpkg.com e.g.:

```html
<script src="https://unpkg.com/mt940js@latest/dist/index.min.js"></script>
<script>
  const parser = new mt940js.Parser();
  ...
</script>
```

## API and usage

Main parser class - `Parser` - parses input text (e.g. read from a file) into array of statements (a file may contain one or more). Each output statement contains a set of attributes, describing opening balance, statement number, etc and also an array of transactions.

**Example**

```javascript
const mt940js = require('mt940js');
const parser  = new mt940js.Parser();

const statements = parser.parse(fs.readFileSync('./some_path', 'utf8'));

for (let s of statements) {
 console.log(s.number.statement, s.statementDate);

 for (let t of s.transactions) {
   console.log(t.amount, t.currency);
 }
}
```

### Statement

-  `transactionReference` {string} - tag 20 reference
-  `relatedReference` {string} - tag 21 reference, *optional*
-  `accountIdentification` {string} - tag 25 own bank account identification
-  `number.statement` {string} - tag 28 main statement number
-  `number.sequence` {string} - tag 28 statement sub number (sequence), *optional*
-  `number.section` {string} - tag 28 statement sub sub number (present on some banks), *optional*
-  `openingBalanceDate` {Date} - tag 60 statement opening date
-  `closingBalanceDate` {Date} - tag 62 statement closing date
-  `closingAvailableBalanceDate` {Date} - tag 64 closing available balance date, default = closing date
-  `forwardAvailableBalanceDate` {Date} - tag 65 forward available balance date, default = closing available date
-  `statementDate` {Date} - abstraction for statement date = `closingBalanceDate`
-  `currency` {string} - statement currency (USD, EUR ...)
-  `openingBalance` {Number} - beginning balance of the statement (with sign, based on debit/credit mark)
-  `closingBalance` {Number} - ending balance of the statement (with sign, based on debit/credit mark)
-  `closingAvailableBalance` {Number} - tag 64 closing available balance, default = closing balance
-  `forwardAvailableBalance` {Number} - tag 65 forward available balance, default = closing available
-  `informationToAccountOwner` {string} - additional statement level information
-  `transactions` {array}  - collection of transactions
-  `messageBlocks` {object} - statement message blocks, if present (EXPERIMENTAL)
 
**Each Transaction contains data of tag 61 (and tag 86 for details)**

- `date` {Date} - transaction date
- `amount` {Number} - transaction amount (with sign, Credit+, Debit-)
- `reversal` {Boolean} - transaction is a reversal
- `currency` {string} - transaction currency (copy of statement currency)
- `details` {string} - content of relevant 86 tag(s), may be multiline (`\n` separated)
- `transactionType` {string} - MT940 transaction type code (e.g. NTRF ...)
- `reference` {string} - payment reference field
- `entryDate` {Date} - entry date field, *optional*
- `fundsCode` {string} - funds code field, *optional*
- `bankReference` {string} - bank reference, *optional*
- `extraDetails` {string} - extra details (supplementary details), *optional*
- `structuredDetails` {Object} - structured details if detected, in for of `{ subtag: value }` e.g. `{ '20': '123456' }`
- `nonSwift` {string} - optional, content of NS tag which happened in the context of transaction (after tags 61 or 86), can be multiline (separated by `\n`)

Each statement is validated for: 

- all strictly required tags
- opening/closing balance currency is the same
- opening balance + turnover = closing balance

### Invocation

The `Parser` has just one method - `parse(data, withTags = false)` - which will convert raw mt940 string to an array of statements described above. The optional `withTags` parameter would preserve parsed tags to an additional `tags` attribute of a statement (for any additional further analysis).

### Support for field 86 structure

Currently the library supports the following tag formats:
- `'<sep>DD'`, where `<sep>` can be `'>'` or `'?'`
- `'/TAG/value'`, where `TAG` is 2 to 4 uppercase chars.

Example:

```
'>20some details >30more data'
or
'/ORDP/Smith Corp'
``` 

The parser attempts to detect if field 86 contains tags like these and, if yes, adds `structuredDetails` attribute to a statement line. Tag digits are not interpreted as they are not standardized among different banks. Parsing 86 structure can be force disabled by passing `{ no86Structure: true }` to the constructor.

```javascript
// let incoming file contain one line with 86 field:
// '>20some details>30more data'

const statements = ... // parsing here

for (let s of statements) {
 for (let t of s.transactions) {
   console.log(t.structuredDetails);
   // { '20': 'some details',
   //   '30': 'more data' }
 }
}
```

### Middlewares

**Currently experimental, may change**

The library support post processing middlewares which is called before returning parsed result. To append a middleware call `usePostParse` passing `fn(statement, next)`. Middlewares are called in the order of appending. Middlewares modify statement object directly. Beware that input data may contain several statements, middlewares are called over each of them one by one.

```javascript
  const parser = new Parser();
  parser.usePostParse((s, next) => {
    s.hasOverdraft = (s.closingBalance < 0);
    next();
  });
```

## Contribution
Contribution is welcomed :)

## TODO
- pre parsing middlewares
- parsing structure of block messages

## Author
[Alexander Tsybulsky](https://github.com/a-fas)

## License
The code is licensed under Apache-2.0 License. Please see [LICENSE](./LICENSE) for details.

## Credits
Inspired by https://github.com/WoLpH/mt940

## Standard references
- https://www2.swift.com/knowledgecentre/publications/us9m_20180720/2.0
- https://www.paiementor.com/payment-messages/
- https://www.paiementor.com/swift-mt-message-structure-blocks-1-to-5/
- https://www.nordea.se/Images/39-16149/MT940-file-description.pdf

changelog.txt
mt940-js changelog
==================

Legend
------
* : fixed
! : changed
+ : added
- : removed

2020-??-?? v1.4.0
------------------
+ umd build support, and unpkg.com as a consequence
* misc: update dependencies, switch to github actions, switch to nyc, fix eslint dependency

2019-12-30 v1.3.4
------------------
* Dependencies update, npm audit fix

2019-12-03 v1.3.3
------------------
+ Support for NS fields in the context of transactions (saved to 'nonSwift' attribute)
+ Support for GVC at the beginning of payment details (thanks to @artur-krueger)

2019-01-31 v1.3.1
------------------
+ no86structure param to skip 86 tag structure detection
* minor fixes and refactorings

2019-01-29 v1.3.0
------------------
+ experimental messageBlocks detection and partial parsing '{1:...}'

2019-01-29 v1.2.1
------------------
* bug fixes: tag 61 extra details, leading spaces in tag 86, statement level informationToAccountOwner (thanks to @nickidk4)
* potential bug fix: tags starts at the line beginning

2019-01-19 v1.2.0
------------------
+ support for reversal flag, field 61 also contains isReversal attribute
+ support for fields 64 and 65 (closingAvailableBalance, forwardAvailableBalance)
+ support for /XXX/ field 86 subtags
+ support post parse middlewares

2016-09-08 v1.1.0
------------------
+ added some support for structured 86 field

2016-12-11 v1.0.0
------------------
First release :tada:


test/tags.js
const assert = require('chai').assert;
const tags = require('../lib/tags');

describe('Tags', () => {
  describe('TagFactory', () => {
    const tf  = new tags.TagFactory();

    it('should create tag 20 (TransactionReferenceNumber)', () => {
      const ref = 'REFERENCE';
      const tag = tf.createTag('20', null, ref);
      assert.equal(ref, tag.fields.transactionReference);
    });

    it('should create tag 21 (RelatedReference)', () => {
      const ref = 'REFERENCE';
      const tag = tf.createTag('21', null, ref);
      assert.equal(ref, tag.fields.relatedReference);
    });

    it('should create tag 25 (AccountIdentification)', () => {
      const account = '123456789';
      const tag = tf.createTag('25', null, account);
      assert.equal(account, tag.fields.accountIdentification);
    });

    it('should create tag 28 (StatementNumber)', () => {
      const str = '998/1';
      const tag = tf.createTag('28', null, str);
      assert.equal('998', tag.fields.statementNumber);
      assert.equal('1', tag.fields.sequenceNumber);
    });

    it('should create tag NS (NonSwift)', () => {
      const str = 'XYZ';
      const tag = tf.createTag('NS', null, str);
      assert.equal('XYZ', tag.fields.nonSwift);
    });

    it('should create tag 60 (OpeningBalance)', () => {
      const str = 'C160507EUR123,89';
      const tag = tf.createTag('60', null, str);
      assert.equal('2016-05-07', tag.fields.date.toISOString().substr(0,10));
      assert.equal('EUR', tag.fields.currency);
      assert.equal(123.89, tag.fields.amount);
    });

    it('should create tag 62 (ClosingBalance)', () => {
      const str = 'C160507EUR123,89';
      const tag = tf.createTag('62', null, str);
      assert.equal('2016-05-07', tag.fields.date.toISOString().substr(0,10));
      assert.equal('EUR', tag.fields.currency);
      assert.equal(123.89, tag.fields.amount);
    });

    it('should create tag 61 (StatementLine)', () => {
      const str = '1605070507D123,89NTRFNONREF//B4E07XM00J000023';
      const tag = tf.createTag('61', null, str);
      assert.equal('2016-05-07', tag.fields.date.toISOString().substr(0,10));
      assert.equal('2016-05-07', tag.fields.entryDate.toISOString().substr(0,10));
      assert.equal(-123.89, tag.fields.amount);
      assert.equal('NTRF', tag.fields.transactionType);
      assert.equal('NONREF', tag.fields.reference);
      assert.equal('B4E07XM00J000023', tag.fields.bankReference);
    });

    it('should create tag 86 (TransactionDetails)', () => {
      const str = 'Some text here';
      const tag = tf.createTag('86', null, str);
      assert.equal('Some text here', tag.fields.transactionDetails);
    });

    it('should create tag with a subId', () => {
      const str = '998/1';
      const tag = tf.createTag('28', 'C', str);
      assert.equal('998', tag.fields.statementNumber);
      assert.equal('1', tag.fields.sequenceNumber);
    });

    it('should throw unknown tag', () => {
      const str = 'Some data';
      assert.throws(
        tf.createTag.bind(tf, 'XX', null, str),
        /Unknown tag/);
    });

    it('should throw wrong content', () => {
      const str = 'Some data';
      assert.throws(
        tf.createTag.bind(tf, '28', null, str),
        /Cannot parse/);
    });

    it('Abstract Tag and TagBalance', () => {
      assert.throws(
        () => {let a = new tags.Tag;},
        /Tag instances/);
      assert.throws(
        () => {let a = new tags.TagBalance('C160507EUR123,89');},
        /TagBalance instances/);
    });

  });
});


test/parser.js
const assert  = require('chai').assert;
const Parser  = require('../lib/parser');
const Tags    = require('../lib/tags');
const helpers = require('../lib/helperModels');
const parser  = new Parser();

function dummyStatementLines() {
  return [
    '{1:XXX0000000000}{2:XXXXXXXXX}{4:',
    ':20:B4E08MS9D00A0009',
    ':21:X',
    ':25:123456789',
    ':28C:123/1',
    ':60F:C140507EUR0,00',
    ':61:1405070507C500,00NTRFNONREF//AUXREF',
    ':86:LINE1',
    'LINE2',
    ':62F:C140508EUR500,00',
    '-}'
  ];
}

function expectedStatement() {
  return {
    transactionReference: 'B4E08MS9D00A0009',
    relatedReference: 'X',
    accountIdentification: '123456789',
    number: {
      statement: '123',
      sequence: '1',
      section: ''
    },
    statementDate: helpers.Date.parse('14', '05', '08'),
    openingBalanceDate: helpers.Date.parse('14', '05', '07'),
    closingBalanceDate: helpers.Date.parse('14', '05', '08'),
    currency: 'EUR',
    openingBalance: 0.0,
    closingBalance: 500.0,
    transactions: [
      {
        amount: 500.00,
        currency: 'EUR',
        reference: 'NONREF',
        bankReference: 'AUXREF',
        transactionType: 'NTRF',
        date: helpers.Date.parse('14', '05', '07'),
        entryDate: helpers.Date.parse('14', '05', '07'),
        details: 'LINE1\nLINE2',
        extraDetails: '',
        fundsCode: '',
      }
    ]};
}

function dummyGroups() {
  return [
    [ // Normal 1
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
      new Tags.TagAccountIdentification('123456789'),
      new Tags.TagStatementNumber('123/1'),
      new Tags.TagOpeningBalance('C140507EUR0,00'),
      new Tags.TagStatementLine('1405070507C500,00NTRFNONREF//AUXREF'),
      new Tags.TagTransactionDetails('DETAILS'),
      new Tags.TagClosingBalance('C140508EUR500,00')
    ],
    [ // Normal 2, 2 detail lines and 2 transactions
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
      new Tags.TagRelatedReference('X'),
      new Tags.TagAccountIdentification('123456789'),
      new Tags.TagStatementNumber('123/1'),
      new Tags.TagOpeningBalance('C140507EUR0,00'),
      new Tags.TagStatementLine('1405070507C500,00NTRFNONREF//AUXREF'),
      new Tags.TagTransactionDetails('LINE1'),
      new Tags.TagTransactionDetails('LINE2'),
      new Tags.TagStatementLine('1405070507C0,00NTRFNONREF2'),
      new Tags.TagTransactionDetails('LINE1'),
      new Tags.TagClosingBalance('C140508EUR500,00')
    ],
    [ // missing tags
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
    ],
    [ // missing tags
      new Tags.TagClosingBalance('C140508EUR500,00')
    ],
    [ // missing tags
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
      new Tags.TagOpeningBalance('C140507EUR0,00'),
      new Tags.TagClosingBalance('C140508EUR500,00')
    ],
    [ // inconsistent currency
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
      new Tags.TagAccountIdentification('123456789'),
      new Tags.TagStatementNumber('123/1'),
      new Tags.TagOpeningBalance('C140507EUR0,00'),
      new Tags.TagStatementLine('1405070507C500,00NTRFNONREF//AUXREF'),
      new Tags.TagTransactionDetails('DETAILS'),
      new Tags.TagClosingBalance('C140508USD500,00')
    ],
    [ // inconsistent balances
      new Tags.TagTransactionReferenceNumber('B4E08MS9D00A0009'),
      new Tags.TagAccountIdentification('123456789'),
      new Tags.TagStatementNumber('123/1'),
      new Tags.TagOpeningBalance('C140507EUR0,00'),
      new Tags.TagStatementLine('1405070507C400,00NTRFNONREF//AUXREF'),
      new Tags.TagTransactionDetails('DETAILS'),
      new Tags.TagClosingBalance('C140508EUR500,00')
    ],
  ]
}

///////////////////////////////////////////////////////////////////////////////
// TESTS
///////////////////////////////////////////////////////////////////////////////

describe('Parser', () => {
  describe('Parser methods', () => {

    it('_splitAndNormalize', () => {
      const result = parser._splitAndNormalize('abc   \r\n   \r\n-');
      assert.deepEqual(result, ['abc']);
    });

    it('_parseLines', () => {
      const result = [...parser._parseLines(dummyStatementLines())];
      assert.equal(8, result.length);
      assert.deepEqual(result[0], {id: '20', subId: '',  data: ['B4E08MS9D00A0009']});
      assert.deepEqual(result[1], {id: '21', subId: '',  data: ['X']});
      assert.deepEqual(result[2], {id: '25', subId: '',  data: ['123456789']});
      assert.deepEqual(result[3], {id: '28', subId: 'C', data: ['123/1']});
      assert.deepEqual(result[4], {id: '60', subId: 'F', data: ['C140507EUR0,00']});
      assert.deepEqual(result[5], {id: '61', subId: '',  data: ['1405070507C500,00NTRFNONREF//AUXREF']});
      assert.deepEqual(result[6], {id: '86', subId: '',  data: ['LINE1', 'LINE2']});
      assert.deepEqual(result[7], {id: '62', subId: 'F', data: ['C140508EUR500,00']});
    });

    it('_groupTags', () => {
      const groups = dummyGroups().slice(0,2);
      const result = parser._groupTags(groups[0].concat(groups[1]));
      assert.deepEqual(result, groups);
    });

    it('_buildStatement', () => {
      const group  = dummyGroups()[1]; // With extra detail line
      const result = parser._buildStatement(group);

      let exp  = expectedStatement();
      // patch
      exp.transactions.push({
        amount: 0.00,
        currency: 'EUR',
        reference: 'NONREF2',
        bankReference: '',
        transactionType: 'NTRF',
        date: helpers.Date.parse('14', '05', '07'),
        entryDate: helpers.Date.parse('14', '05', '07'),
        details: 'LINE1',
        extraDetails: '',
        fundsCode: '',
      });
      assert.deepEqual(result, exp);
    });

    it('_validateGroup', () => {
      const groups = dummyGroups();
      assert.throws(parser._validateGroup.bind(parser, groups[2]), /Mandatory tag/);
      assert.throws(parser._validateGroup.bind(parser, groups[3]), /Mandatory tag/);
      assert.throws(parser._validateGroup.bind(parser, groups[4]), /Mandatory tag/);
      assert.throws(parser._validateGroup.bind(parser, groups[5]), /Currency markers/);
      assert.throws(parser._validateGroup.bind(parser, groups[6]), /Sum of lines/);
    });

  });

  describe('Integration test', () => {

    it('should pass', () => {
      const result = parser.parse(dummyStatementLines().join('\n'));
      assert.equal(result.length, 1);
      assert.deepEqual(result[0], expectedStatement());
    });

  });
});


test/helpers.js
const assert = require('chai').assert;
const helpers = require('../lib/helperModels');

describe('Helpers', () => {
  describe('Bank date parser', () => {
    it('should work at all (with full date)', () => {
      assert.equal('2016-12-01', helpers.Date.parse(2016,12,1).toISOString().substr(0, 10));
      assert.equal('1996-12-01', helpers.Date.parse(1996,12,1).toISOString().substr(0, 10));
    });

    it('should work with YY (2-digit) year', () => {
      assert.equal('2016-12-01', helpers.Date.parse(16,12,1).toISOString().substr(0, 10));
    });
  });

  describe('Bank amount', () => {
    describe('Parse', () => {
      it('should parse debit amount -> negative', () => {
        assert.equal(helpers.Amount.parse('D', '123.34'), -123.34);
      });

      it('should parse credit amount -> positive', () => {
        assert.equal(helpers.Amount.parse('C', '123.34'), 123.34);
      });

      it('should parse amount with ,', () => {
        assert.equal(helpers.Amount.parse('C', '123,34'), 123.34);
      });

      it('should fail if wrong indicator passed', () => {
        assert.throws(helpers.Amount.parse.bind(null, 'X', '123,34'));
      });

      it('should fail if wrong amount passed', () => {
        assert.throws(helpers.Amount.parse.bind(null, 'D', 'XXXXXX'));
      });
    });

    describe('isEqual', () => {
      it('Basic checks', () => {
        assert.isTrue(helpers.Amount.isEqual(123.23, 123.2301));
        assert.isFalse(helpers.Amount.isEqual(123.23, 123.235));
      });
    });
  });


});


cli.js
const fs          = require('fs');
const MT940Parser = require('./lib/parser');

function banner() {
  console.log('Usage: node cli.js <filename>');
  console.log('--');
  console.log('  filename - mt940 text file in UTF8 encoding');
}

const argv = process.argv.slice(2);
let files  = [];

if (!argv.length) {
  banner();
  process.exit(1);
}

argv.forEach(arg => {
  switch (arg) {
  case '--help':
  case '-h':
    banner();
    process.exit(1);
    break;
  default:
    try {
      fs.accessSync(arg);
      files.push(arg);
    } catch (e) {
      console.error(`Cannot access file: ${arg}`);
      process.exit(1);
    }
  }
});

files.forEach(file => {
  const parser     = new MT940Parser();
  const data       = fs.readFileSync(file, { encoding: 'utf8' });
  const statements = parser.parse(data);

  statements.forEach(stmt => {
    console.log('--');
    console.log(JSON.stringify(stmt, null, '  '));
  });
});


typings/index.d.ts
// Type definitions for mt940js
// BETA !, PRs welcomed !

declare namespace mt940js {
    declare class BankDate {
        static parse(year: string, month: string, day: string): Date;
    }

    declare enum DCMark {
        D = 'D',
        C = 'C',
        RD = 'RD',
        RC = 'RC',
    }

    interface StringMap { [key: string]: string }
    type StructuredField86 = StringMap;

    declare class BankAmount {
        static parse(dcmark: DCMark, amountStr: string): number;
        static isEqual(a: number, b: number): boolean;
    }

    type TagID = string | number;
    declare class Tag {
        static get ID(): TagID;
        static get PATTERN(): RegExp;
        constructor(data: string);
        data: string;
        fields: StringMap;
    }

    interface StatementNumber {
        statement: string;
        sequence?: string;
        section?: string;
    }
    interface StatementTransaction {
        date: Date;
        amount: number;
        isReversal: boolean;
        currency: string;
        details: string;
        transactionType: string;
        reference: string;
        entryDate?: Date;
        fundsCode?: string;
        bankReference?: string;
        extraDetails?: string;
        structuredDetails?: StructuredField86;
        nonSwift?: string;
    }

    interface Statement {
        transactionReference: string;
        relatedReference: string;
        accountIdentification: string;
        number: StatementNumber;
        openingBalanceDate: Date;
        closingBalanceDate: Date;
        closingAvailableBalanceDate: Date;
        forwardAvailableBalanceDate: Date;
        statementDate: Date;
        currency: string
        openingBalance: number;
        closingBalance: number;
        closingAvailableBalance: number;
        forwardAvailableBalance: number;
        informationToAccountOwner: string;
        messageBlocks: any; // EXPERIMENTAL
        transactions: StatementTransaction[];
    }
    interface NextFunction {
        (): void;
    }
    interface postParseMiddlewareCb {
        (statement: Statement, next: NextFunction): Statement;
    }

    interface ParserParams {
        no86Structure ?: boolean;
    }
    declare class Parser {
        constructor(params: ParserParams);
        parse(data: string, withTags: boolean = false): Statement[];
        usePostParse(fn: postParseMiddlewareCb): void;
    }

}


lib/helperModels.js
/**
 * Helper models
 * @module lib/helperModels
 */

/** helper class to parse date based on 2-digit year */
class BankDate {
  /**
   * Parses date
   * @param {string} year - 4-digit or 2-digit year, 20xx assumed for the latter
   * @param {string} month - month number (starting from 1 = January)
   * @param {string} day - day number
   * @return {Date} a Date object
   * @static
   */
  static parse(year, month, day) {
    let fullyear = Number.parseInt(year, 10);
    if (fullyear < 100) {
      fullyear += 2000;
    }
    return new Date(Date.UTC(fullyear, Number.parseInt(month, 10) - 1, day));
  }
}

/** helper class to parse amount and identify it's sign based on Debit/Credit mark */
class BankAmount {

  /**
   * Parses amount, identifies sign based on Debit or Credit mark.
   * @param {string} dcmark - D or C = Debit or credit (C = positive), prefix 'R' -> change of sign, reversed
   * @param {string} amoutStr - string with positive float number
   * @return {float} amount, rounded to 2 fractional digits, with sign
   * @static
   */
  static parse(dcmark, amountStr) {
    let rev, dc = dcmark;
    if (dcmark.length === 2) {
      rev = dcmark[0];
      dc = dcmark[1];
    }
    if (rev && rev !== 'R') throw Error( `Wrong reversal mark: ${dcmark}` );
    if (! (dc === 'D' || dc === 'C') ) throw Error( `Wrong debit/credit mark: ${dcmark}` );

    let amount = Number.parseFloat(amountStr.replace(',', '.'));
    if (Number.isNaN(amount)) { throw Error( `Amount cannot be parsed: ${amountStr}` ) }
    if (amount < 0) { throw Error( `Positive amount string expected: ${amountStr}` ) }

    if (dc === 'D') amount = -amount; // Bank debit = minus
    if (rev) amount = -amount;
    amount = Math.round(amount * 100) / 100;
    return amount;
  }

  /**
   * Compares amounts (floats in fact), returns True if difference < 0.001
   * @param {float} a - number 1
   * @param {float} b - number 2
   * @static
   */
  static isEqual(a, b) {
    return Math.abs(a - b) < 0.001;
  }
}

module.exports = {
  /** Bank amount, parses debit/credit mark + amount string */
  Amount: BankAmount,

  /** Bank date, parses date with 2digit year */
  Date: BankDate
};


lib/field86structure.js
class Field86StructureParser {
  static buildTagRe(details) {
    const prefix  = details.charAt();

    // check first symbol is known separator
    let tagRe;
    if (!prefix) {
      return;
    } else if (prefix === '/') {                // assume /XXX/ fields
      tagRe = '\\/[0-9A-Z]{2,4}\\/';
    } else if ('>?'.includes(prefix)) {  // assume >DD fields
      tagRe = `\\${prefix}\\d{2}`;
    } else return; // known separator not found

    return {
      detect: new RegExp(`^${tagRe}`),    // line must begin with a tag
      match:  new RegExp(`(${tagRe})(.*?)(?=(${tagRe})|$)`, 'g'), // matcher
      // item:   new RegExp(`^(${tagRe})(.*)`),
      strip:  new RegExp(`\\${prefix}`, 'g'),
    };
  }

  /**
   * Detects if field 86 is structured and attempts to parse it
   */
  static parse(details) {
    details = details.replace(/\n/g, '').trim();

    const parsedStruc = {};

    if (details.match(/^\d\d\d[?>]/)) {
      parsedStruc.gvc = details.substr(0,3);
      details = details.substr(3);
    }

    const rule = Field86StructureParser.buildTagRe(details);
    if (!rule) return;
    if (!rule.detect.test(details)) return; // string must start with tag
    let match = rule.match.exec(details);
    while (match) {
      const subTag = match[1].replace(rule.strip, '');
      parsedStruc[subTag] = match[2];
      match = rule.match.exec(details);
    }

    return parsedStruc;
  }
}

module.exports = Field86StructureParser;


lib/tags.js
/**
 * Tags parsing
 * @module lib/tags
 */

const helperModels = require('./helperModels');

/**
 * Builds a proper tag based on id, subid and tag content
 */
class TagFactory {
  /** constructor - Builds tag map */
  constructor () {
    this.tagMap = [
      TagTransactionReferenceNumber,
      TagRelatedReference,
      TagAccountIdentification,
      TagStatementNumber,
      TagNonSwift,
      TagOpeningBalance,
      TagClosingBalance,
      TagStatementLine,
      TagTransactionDetails,
      TagClosingAvailableBalance,
      TagForwardAvailableBalance,
      TagMessageBlock,
    ].reduce((map, i) => map.set(i.ID, i), new Map());
  }

  /**
  * Create proper tag instance
  * @param {string} id - tag id (21, 61 ...)
  * @param {string} subId - optional tag sub Id (C, F ...)
  * @param {string} data - tag content (without id part)
  * @return {Tag} new Tag (specific subclass) instance
  */
  createTag(id, subId, data) {
    const tagId     = isNaN(id) ? id : Number.parseInt(id, 10);
    const fullTagId = tagId.toString() + (subId ? subId.toString() : '');
    const tagClass  = this.tagMap.get(fullTagId) || this.tagMap.get(tagId);

    if (!tagClass) { throw new Error(`Unknown tag ${fullTagId}`) }

    return new tagClass(data);
  }
}


/**
* Abstract Tag base class
* @abstract
*/
class Tag {
  /** Tag ID */
  static get ID() { return 0 }
  /** Tag data regex pattern */
  static get PATTERN() { return /^(.*)/ }

  constructor(data) {
    if (new.target === Tag) {
      throw new TypeError('Cannot construct Tag instances directly');
    }
    this.id   = this.constructor.ID;
    this.re   = this.constructor.PATTERN;
    this.data = data;
    this._parse();
  }

  /**
   * parses tag data nad saves it as `fields` attribute
   * @private
   */
  _parse() {
    const match = this._nextMatch();
    if (!match) { throw Error(`Cannot parse tag ${this.id}: ${this.data}`) }
    this.fields = this._extractFields(match);
  }

  _nextMatch() {
    return this.re.exec(this.data);
  }

  _extractFields(match) { return {} } // eslint-disable-line no-unused-vars
}

class TagTransactionReferenceNumber extends Tag {
  static get ID() { return 20 }
  static get PATTERN() { return /^(.{0,16})/ }
  _extractFields(match) {
    return {
      transactionReference: match[1]
    };
  }
}

class TagRelatedReference extends Tag {
  static get ID() { return 21 }
  static get PATTERN() { return /^(.{0,16})/ }
  _extractFields(match) {
    return {
      relatedReference: match[1]
    };
  }
}

class TagAccountIdentification extends Tag {
  static get ID() { return 25 }
  static get PATTERN() { return /^(.{0,35})/ }
  _extractFields(match) {
    return {
      accountIdentification: match[1]
    };
  }
}

class TagStatementNumber extends Tag {
  static get ID() { return 28 }
  static get PATTERN() { return /^(\d{1,5})(\/(\d{1,5}))?(\/(\d{1,5}))?/ }
  _extractFields(match) {
    return {
      statementNumber: match[1],
      sequenceNumber:  match[3] || '',
      sectionNumber:   match[5] || ''
    };
  }
}

class TagNonSwift extends Tag {
  static get ID() { return 'NS' }
  static get PATTERN() { return /^(.*)/ }
  _extractFields(match) {
    return {
      nonSwift: match[1]
    };
  }
}

class TagBalance extends Tag {
  static get PATTERN() {
    const re = '^([DC])'                  // DC indicator
             + '(\\d{2})(\\d{2})(\\d{2})' // Date
             + '([A-Z]{3})'               // Currency
             + '([0-9,]{0,16})';          // Amount
    return new RegExp(re);
  }
  constructor(data) {

    super(data);
    if (new.target === TagBalance) {
      throw new TypeError('Cannot construct TagBalance instances directly');
    }
  }
  _extractFields(match) {
    return {
      date: helperModels.Date.parse(match[2], match[3], match[4]),
      currency: match[5],
      amount: helperModels.Amount.parse(match[1], match[6])
    };
  }
}

class TagOpeningBalance extends TagBalance { static get ID() { return 60 } }
class TagClosingBalance extends TagBalance { static get ID() { return 62 } }
class TagClosingAvailableBalance extends TagBalance { static get ID() { return 64 } }
class TagForwardAvailableBalance extends TagBalance { static get ID() { return 65 } }

class TagStatementLine extends Tag {
  static get ID() { return 61 }
  static get PATTERN() {
    const re = '^(\\d{2})(\\d{2})(\\d{2})' // Date
             + '((\\d{2})(\\d{2}))?'      // Entry date
             + '(R?[DC])([A-Z])?'         // DC indicator + funds code
             + '([0-9,]{0,16})'           // Amount
             + '([A-Z][A-Z0-9]{3})'       // Transaction type
             + '([^/\\n]{0,16})'          // Customer reference
             + '(//(.{0,16}))?'           // Bank reference
             + '(\\n(.{0,34}))?';         // Extra
    return new RegExp(re);
  }
  _extractFields(match) {
    return {
      date:            helperModels.Date.parse(match[1], match[2], match[3]),
      entryDate:       match[4] && helperModels.Date.parse(match[1], match[5], match[6]) || '',
      fundsCode:       match[8] || '',
      amount:          helperModels.Amount.parse(match[7], match[9]),
      isReversal:      match[7][0] === 'R',
      transactionType: match[10],
      reference:       match[11],
      bankReference:   match[13] || '',
      extraDetails:    match[15] || ''
    };
  }
}

class TagTransactionDetails extends Tag {
  static get ID() { return 86 }
  static get PATTERN() { return /^([\s\S]{0,390})/ }
  _extractFields(match) {
    return {
      transactionDetails: match[1]
    };
  }
}

class TagMessageBlock extends Tag {
  static get ID() { return 'MB' }
  static get PATTERN() {
    const re =
      '(^-\\})|('
      + '\\{(\\d):'
      + '(.*?)'
      + '($|\\}(?=(\\{\\d:)|$))'
      + ')';
    return new RegExp(re, 'g');
  }
  _extractFields(match) {
    const fields = {};
    while (match) {
      if (match[0] === '-}') {
        fields['EOB'] = '';
      } else {
        fields[match[3]] = match[4];
      }
      match = this._nextMatch();
    }
    this.isStarting = (fields['1'] !== undefined); // has message 1
    return fields;
  }
}

module.exports = {
  TagFactory,
  Tag,
  TagTransactionReferenceNumber,
  TagRelatedReference,
  TagAccountIdentification,
  TagStatementNumber,
  TagNonSwift,
  TagBalance,
  TagOpeningBalance,
  TagClosingBalance,
  TagStatementLine,
  TagTransactionDetails,
  TagClosingAvailableBalance,
  TagForwardAvailableBalance,
  TagMessageBlock,
};
